runtime/shibuya/src/xcm_config.rs

```diff
@@ -260,3 +271,72 @@ impl cumulus_pallet_dmp_queue::Config for Runtime {
     type XcmExecutor = XcmExecutor<XcmConfig>;
     type ExecuteOverweightOrigin = EnsureRoot<AccountId>;
 }
+
+/// Convert `AccountId` to `MultiLocation`.
+pub struct AccountIdToMultiLocation;
+impl Convert<AccountId, MultiLocation> for AccountIdToMultiLocation {
+    fn convert(account: AccountId) -> MultiLocation {
+        X1(AccountId32 {
+            network: None,
+            id: account.into(),
+        })
+        .into()
+    }
+}
+
+parameter_types! {
+    /// The absolute location in perspective of the whole network.
+    pub ShibuyaLocationAbsolute: MultiLocation = MultiLocation {
+        parents: 1,
+        interior: X1(
+            Parachain(ParachainInfo::parachain_id().into())
+        )
+    };
+    /// Max asset types for one cross-chain transfer. `2` covers all current use cases.
+    /// Can be updated with extra test cases in the future if needed.
+    pub const MaxAssetsForTransfer: usize = 2;
+}
+
+/// Convert `AssetId` to optional `MultiLocation`. The impl is a wrapper
+/// on `ShibuyaAssetLocationIdConverter`.
+pub struct AssetIdConvert;
+impl Convert<AssetId, Option<MultiLocation>> for AssetIdConvert {
+    fn convert(asset_id: AssetId) -> Option<MultiLocation> {
+        ShibuyaAssetLocationIdConverter::reverse_ref(&asset_id).ok()
+    }
+}
+
+/// `MultiAsset` reserve location provider. It's based on `RelativeReserveProvider` and in
+/// addition will convert self absolute location to relative location.
+pub struct AbsoluteAndRelativeReserveProvider<AbsoluteLocation>(PhantomData<AbsoluteLocation>);
+impl<AbsoluteLocation: Get<MultiLocation>> Reserve
+    for AbsoluteAndRelativeReserveProvider<AbsoluteLocation>
+{
+    fn reserve(asset: &MultiAsset) -> Option<MultiLocation> {
+        RelativeReserveProvider::reserve(asset).map(|reserve_location| {
+            if reserve_location == AbsoluteLocation::get() {
+                MultiLocation::here()
+            } else {
+                reserve_location
+            }
+        })
+    }
+}
+
+impl orml_xtokens::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type Balance = Balance;
+    type CurrencyId = AssetId;
+    type CurrencyIdConvert = AssetIdConvert;
+    type AccountIdToMultiLocation = AccountIdToMultiLocation;
+    type SelfLocation = ShibuyaLocation;
+    type XcmExecutor = XcmExecutor<XcmConfig>;
+    type Weigher = Weigher;
+    type BaseXcmWeight = UnitWeightCost;
+    type UniversalLocation = UniversalLocation;
+    type MaxAssetsForTransfer = MaxAssetsForTransfer;
+    // Default impl. Refer to `orml-xtokens` docs for more details.
+    type MinXcmFee = DisabledParachainFee;
```

Comment:
If I understand correctly, this will disable _double XCM support_ (not sure how it's properly called :slightly_smiling_face: ), e.g. **Astar -> Polkadot -> Acala**?

Reply:
If you mean sending assets from one chain to another with a hop through the reserve chain, like sending DOT from Astar to Acala, or sending ASTR from Moonbeam to Acala, this default impl actually works.😸

The non-default `MinXcmFee` is for an edge case that sends assets to its reserve chain which only accepts relay chain assets as fees. In this case two XCM message needs to be sent, one to relay chain and the other to the dest reserve chain. And the `MinXcmFee` serves as an indicator of how to break fees for these two messages. The only real-life case I know was a while ago when Statemine only accepts KSM as XCM fees, and Acala try to support sending RMRK to it. You may find more details [here](https://github.com/open-web3-stack/open-runtime-module-library/tree/master/xtokens#transfer-multiple-currencies), https://github.com/open-web3-stack/open-runtime-module-library/issues/698 . (P.S. now statemine/t support other assets as fees with https://github.com/paritytech/cumulus/pull/1278)

Reply:
I see, thanks for the explanation! :pray: